Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h

```diff
@@ -63,6 +68,13 @@ class WebPermissionController final : public CanMakeWeakPtr<WebPermissionControl
         CompletionHandler<void(std::optional<WebCore::PermissionState>)> completionHandler;
         bool isWaitingForReply { false };
     };
+    struct ObserverData {
+        WebCore::ClientOrigin origin;
+        WebCore::PermissionDescriptor descriptor;
+        WebCore::ScriptExecutionContextIdentifier contextIdentifier;
+    };
+
+    HashMap<WebCore::PermissionStatusIdentifier, ObserverData> m_observers;
```

Comment:
m_observers holds WebCore::PermissionObserver because PermissionObserver is not necessarily PermissionStatus by design; it can be any object that needs permission update (for example maybe Notification.permission). This has changed the design. You can avoid this change by making PermissionStatus in worker hold a mainthread PermissionObserver and register that observer.

On the other hand, if we want to make PermissionStatus the only observer, I don't think we need ObserverData or m_observers; just store the data in PermissionStatus and iterate all PermissionsStatus when state has changed.

Reply:
> On the other hand, if we want to make PermissionStatus the only observer, I don't think we need ObserverData or m_observers; just store the data in PermissionStatus and iterate all PermissionsStatus when state has changed.

I recommended against this because it would be very inefficient to dispatch to another thread for each permissionStatus object (doing isolated copy of the origin each time), just to notice that this status object doesn't care about the change notification at all. The current code is way more efficient.

With regards to using a PermissionObserver, I don't have a strict opinion. The proposed code works and is simple, but it is true that we lost some functionality. It is hard for me to evaluate if this functionality is really helpful at all since this is mostly unused code at the moment. We have all this change logic implemented (and that we need to keep maintaining) even though we don't support the change event yet.

Reply:
> I recommended against this because it would be very inefficient to dispatch to another thread for each permissionStatus object (doing isolated copy of the origin each time)
Why would we do copy each time... the PermissionStatus can get the origin from ScriptExecutionContext... we only need to save a copy in PermissionStatus for mainthread use (for searching). This proposal is basically the same as the implementation but without adding another map in PermissionController.

My point is this implementation leaves us a weird state that PermissionObserver is not used, but we make PermissionStatus subclass it and keep a map named m_observers. Either we keep the original design that PermissionObserver is used (and we will extend its usage); or we just remove PermissionObserver. I prefer the first.

Reply:
Ok, I don't mind keeping PermissionObserver and making it work as long as you help Rupin implement it :P

Reply:
Ok, so we decided to only support the observer onchange behavior for window objects in this patch. We'll do another patch later to add support for workers. The changes made to observer earlier in the history of this patch have been reverted. Fixed!